### ISSUE TITLE
Revisit jar version numbering

### ISSUE THREAD

**Issue by mvglasow** [jitpack/jitpack.io#4438]
**Is your feature request related to a problem? Please describe.**
JitPack versioning is very different from standard Maven versioning, as used by the other repositories out there. Standard versioning uses the `<version>` tag in `pom.xml`, while JitPack uses git references (branch names, tags or hash codes).

First of all, that creates difficulties when making jars available via JitPack as well as other repos.

For released versions, this can be alleviated by giving that version a git tag equal to the `<version>` tag in `pom.xml`. However, that may lead to anomalies, as there is no inherent way to prevent `pom.xml` from stating version `0.4.3` in a release tagged `3.7.0` in git.

For unstable versions, the Maven convention seems to be using the upcoming version number with the `-SNAPSHOT` prefix. That is even harder to replicate in JitPack: projects often develop on the main (usually `master`) branch, or a designated development (`dev` or `unstable`) branch, or a feature branch. One could work around this by developing the upcoming version on a branch named after the upcoming version (version `3.7.2` would get developed on a branch named `3.7.2`), but if the official release for that version were to get tagged with the version number, that would create a name conflict between the branch and the tag.

**Describe the solution you'd like**
What I would like is to be able to publish my jar through JitPack as well as other channels (Maven Central, GitLab package registry etc.) and have the same version string everywhere. I understand that, due to the way JitPack works, this is a somewhat challenging task, so I’m putting it out there for discussion. If there is a solution approach to this already, a pointer (even better an FAQ entry) would be very much appreciated.

Once JitPack builds a jar, it would have the version number and could, for future builds, make the jar available under that version number. On the other hand, that version needs to get built first, and if multiple git releases with the same version number get built, that conflict needs to be resolved somehow.

**Describe alternatives you've considered**
See above—basically ensuring the version number corresponds to some git ref, but that comes with some challenges and pitfalls of its own.

**Additional context**
None so far.

**Comment by jitpack-io** [comment]
Hi,

Thank you for starting this discussion!
The most important aspect of JitPack versioning is the ability to trace a release back to a tag/commit. In that sense we have to ignore the version in pom and use the release tag. As you say, this is usually sufficient.

We also have support for snapshots with version `branch-SNAPSHOT`. 
This resolves to a special version `master-1.1-gf129fbd-10`which is derived from `git describe`.
Basically it tells you the commit hash, the last tag (1.1) and number of commits after the last tag (10).

https://jitpack.io/docs/

**Comment by mvglasow** [comment]
In the meantime I have seen that JitPack periodically checks the git repo for updates. For a version number reference, that would alleviate the “first build” problem—the version number can be requested (and JitPack can resolve it) as soon as JitPack has picked it up during one of its routine scans—or, of course, as soon as a version with that code has been referenced by any other means and built by JitPack.

That leaves us with the (im)mutability issue. Here I would suggest:
* Version numbers ending in `-SNAPSHOT` are mutable, and the latest known version will be served.
* Other version numbers become frozen (immutable) after seven days, much like git refs do.

We would still need a strategy to choose between different branches if ambiguities arise. Suggestions here:
* For snapshot versions, preference is given to the `dev` branch (if a branch by that name exists on the repo), then `master`
* Stable versions are taken from `master` only
* If the maintainer has an account, they could be provided with a setting to change these preferences to any branch name of their choice
